### PR TITLE
Change logic to scale based on average messages per pod.

### PR DESCRIPTION
**Pull request by nmohoric**
This changes the logic to use average number of visible messages per pod (rather than total visible messages), immediately autoscales to expected required pod count, and some basic tests around this functionality.

### CODE REVIEW

File: sqs/sqs.py [hunk 1]
```diff
@@ -33,53 +35,63 @@ def message_count(self):
         )
         return int(response['Attributes']['ApproximateNumberOfMessages'])
 
+    def replicas(self):
```
Comment (jingles) [inline sqs/sqs.py:38]: could just make this a `@property`, also could

`return self.deployment().spec.replicas`

File: sqs/sqs.py [hunk 2]
```diff
@@ -33,53 +35,63 @@ def message_count(self):
...
+        deployment = self.deployment()
+        return deployment.spec.replicas
+
     def poll(self):
         message_count = self.message_count()
+        replicas = self.replicas()
+        average_messages = int(message_count / replicas)
```
Comment (jingles) [inline sqs/sqs.py:45]: if `replicas` is ever scaled down to 0 could we hit a `ZeroDivisionError` here?

File: sqs/sqs.py [hunk 3]
```diff
@@ -33,53 +35,63 @@ def message_count(self):
...
         t = time()
-        if message_count >= self.options.scale_up_messages:
+        if average_messages >= self.options.scale_up_messages:
             if t - self.last_scale_up_time > self.options.scale_up_cool_down:
-                self.scale_up()
+                pods = int(ceil(message_count / self.options.scale_up_messages))
```
Comment (jingles) [inline sqs/sqs.py:49]: ditto what happens if there have never been any scale up messages, or perhaps i don't understand where messages come from

File: sqs/sqs.py [hunk 4]
```diff
@@ -33,53 +35,63 @@ def message_count(self):
...
+                self.scale_up(pods)
                 self.last_scale_up_time = t
             else:
                 logger.debug("Waiting for scale up cooldown")
-        if message_count <= self.options.scale_down_messages:
+        if average_messages <= self.options.scale_down_messages:
             if t - self.last_scale_down_time > self.options.scale_down_cool_down:
-                self.scale_down()
+                pods = int(ceil(message_count / self.options.scale_down_messages))
```
Comment (jingles) [inline sqs/sqs.py:56]: and here

File: sqs/sqs.py [hunk 5]
```diff
@@ -33,53 +35,63 @@ def message_count(self):
...
     def deployment(self):
+        if self._deployment:
+            return self._deployment
+
         logger.debug("loading deployment: {} from namespace: {}".format(
             self.options.kubernetes_deployment, self.options.kubernetes_namespace))
         deployments = self.apps_v1.list_namespaced_deployment(
             self.options.kubernetes_namespace, label_selector="app={}".format(self.options.kubernetes_deployment))
-        return deployments.items[0]
+        self._deployment = deployments.items[0]
```
Comment (jingles) [inline sqs/sqs.py:93]: should `update_deployment` set `self._deployment` as well?

File: tests/test_sqs.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    (1, 200, 2),  # Should hit min,
+    (45, 1, 10),  # Recently scaled should stay
+    (250, 200, 13),  # Should scale up to specific number
+    (40, 200, 8),  # Should scale down to specific number
+])
+def test_poll_scale_up(messages, offset, expected, patched, options):
+    client = SQSPoller(options)
+    client.message_count.return_value = messages
+    client.last_scale_up_time = client.last_scale_down_time = time() - offset
+    client.poll()
```
Comment (jingles) [inline tests/test_sqs.py:71]: so `poll()` actually just runs one iteration, and `run()` loops `poll()`?   just seeing if i understand

File: sqs/sqs.py (hunk 2 above)
Comment (nmohoric) [reply sqs/sqs.py:45]: Good catch, will just assume that if it was scaled to 0 then it was done on purpose and shouldn't be scaled (the same way hpa works)

File: sqs/sqs.py (hunk 3 above)
Comment (nmohoric) [reply sqs/sqs.py:49]: Scale up messages is a passed in value from the command line so shouldn't ever be zero.

File: sqs/sqs.py (hunk 4 above)
Comment (nmohoric) [reply sqs/sqs.py:56]: Same here, passed into cli.

File: tests/test_sqs.py (hunk 6 above)
Comment (nmohoric) [reply tests/test_sqs.py:71]: Correct. The joys of working on open source stuff.

**Pull request merged**